### PR TITLE
Device Credentials Installer

### PR DESCRIPTION
**Pull request by plskeggs**
Add serial-based device credentials installer.

Update README.md.

### CODE REVIEW

File: python/modem-firmware-1.3+/create_device_credentials.py [hunk 1]
```diff
@@ -150,6 +164,8 @@ def main():
         csr.sign(priv_key, 'sha256')
 
     # create a device cert
+    device_cert = create_device_cert(args.dv, csr, pub_key, ca_cert, ca_key)
```
Comment (jayteemo) [inline python/modem-firmware-1.3+/create_device_credentials.py:167]: you can remove the stuff below which has been moved into the function

**Review by jayteemo — CHANGES_REQUESTED** [review]
did you forget to add the new py file?

File: python/modem-firmware-1.3+/create_device_credentials.py (hunk 1 above)
Comment (plskeggs) [reply python/modem-firmware-1.3+/create_device_credentials.py:167]: Good grief.  Yes.

File: python/modem-firmware-1.3+/device_credentials_installer.py [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+
+def send_style(line):
+    return ensure_lf(Fore.BLUE + line
+                     + Style.RESET_ALL) if not plain else line
+
+def error_style(line):
+    return ensure_lf(Fore.RED + line + Style.RESET_ALL) if not plain else line
+
+def ask_for_port(selected_port, list_all):
+    """\
```
Comment (jayteemo) [inline python/modem-firmware-1.3+/device_credentials_installer.py]: stray `\`

File: python/modem-firmware-1.3+/device_credentials_installer.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+        print(local_style('deleting sectag {}...'.format(args.sectag)))
+        write_line('AT%CMNG=3,{},0'.format(args.sectag))
+        wait_for_prompt(b'OK', b'ERROR')
+        write_line('AT%CMNG=3,{},1'.format(args.sectag))
+        wait_for_prompt(b'OK', b'ERROR')
+        write_line('AT%CMNG=3,{},2'.format(args.sectag))
+        wait_for_prompt(b'OK', b'ERROR')
+
+    # now get a new certificate signing request (CSR)
+    print(local_style('requesting CSR for sectag {}...'.format(args.sectag)))
```
Comment (jayteemo) [inline python/modem-firmware-1.3+/device_credentials_installer.py]: more specifically, this is something like, "generating private key and requesting a CSR"

File: python/modem-firmware-1.3+/device_credentials_installer.py [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+    write_line('AT%KEYGEN={},2,0'.format(args.sectag))
+    retval, output = wait_for_prompt(b'OK', b'ERROR', store=b'%KEYGEN:')
+    if not retval:
+        print(error_style('unable to create CSR; does it already exist for this sectag?'))
```
Comment (jayteemo) [inline python/modem-firmware-1.3+/device_credentials_installer.py]: "unable to generate private key"

File: python/modem-firmware-1.3+/device_credentials_installer.py [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+        raise RuntimeError("Error loading PEM file " + csr_pem_filepath)
+
+    # check if we have all we need to proceed
+    if len(args.ca) == 0 or len(args.ca_key) == 0:
+        print(local_style('no ca or ca_key provided; skipping creating dev cert'))
+        cleanup()
+        sys.exit(0)
+
+    # load the user's certificate authority (CA)
+    print(local_style('loading ca and key...'))
```
Comment (jayteemo) [inline python/modem-firmware-1.3+/device_credentials_installer.py]: `CA`

File: python/modem-firmware-1.3+/device_credentials_installer.py [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+        answer = ' '
+        while answer not in 'yan':
+            answer = input('--- file {} exists; overwrite, append, or quit (y,a,n)? '.format(csv_filename))
+        if answer == 'n':
+            print(local_style('file will not be overwritten'))
+            return
+        if answer == 'a':
+            mode = 'a'
+
+    with open(csv_filename, mode, newline='\n') as csvfile:
```
Comment (jayteemo) [inline python/modem-firmware-1.3+/device_credentials_installer.py]: would adding exception handling provide any value here?

**Review by jayteemo — CHANGES_REQUESTED** [review]
looks good.
i'm a python novice so i don't really have any code-related comments.
just a couple general cleanup items for log messages:
use upper case "CSV" for printed messages where appropriate.
capitalize first word of all printed messages.

**Comment by plskeggs** [comment]
OK, made the changes.

File: python/modem-firmware-1.3+/device_credentials_installer.py [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+    # for gateways, get to the AT command prompt first
+    if is_gateway:
+        print(local_style('Getting to prompt...'))
+        handle_login()
+
+        print(local_style('Disabling logs...'))
+        write_line('log disable')
+        wait_for_prompt()
+
+        print(local_style('Getting to at mode...'))
```
Comment (jayteemo) [inline python/modem-firmware-1.3+/device_credentials_installer.py]: `AT mode`

File: python/modem-firmware-1.3+/device_credentials_installer.py [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+    modem_dev = str(dev, encoding=full_encoding)
+    if is_gateway:
+        modem_dev = modem_dev.replace("\n", "\\n")
+    write_line('AT%CMNG=0,{},1,"{}"'.format(args.sectag, modem_dev))
+    wait_for_prompt(b'OK', b'ERROR')
+    time.sleep(1)
+
+    # write provisioning information to csv if requested by user
+    if len(args.csv) > 0:
+        print(local_style('{} Provisioning endpoint csv file {}...'
```
Comment (jayteemo) [inline python/modem-firmware-1.3+/device_credentials_installer.py]: `CSV file`

**Review by jayteemo — APPROVED** [review]

**Comment by plskeggs** [comment]
Do you want me to merge this?

**Comment by jayteemo** [comment]
> Do you want me to merge this?

go for it

**Pull request merged**